sharktank/sharktank/utils/load_llm.py

```diff
@@ -0,0 +1,222 @@
+        if self.model.cache.is_direct:
+            return
+        self.free_pages.append(index)
+
+
+class Batch:
+    def __init__(
+        self,
+        parent: TorchGenerator,
+        token_ids: torch.Tensor,
+        seq_lens: torch.Tensor,
+        cache_state: list[torch.Tensor],
+    ):
+        self.bs = token_ids.shape[0]
+        # assert seq_lens.shape[0] == self.bs
+        self.parent = parent
+        self.token_ids = token_ids
+        self.seq_lens = seq_lens
+        self.cache_state = cache_state
+        self.results: list[list[int]] = [[] for _ in range(self.bs)]
+        self.done_result_indices: set[int] = set()
+
+        # Assemble the batch.
+        seq_stride = self.parent.block_seq_stride
+        self.seq_block_ids: list[list[int]] = []
+        for seq_len in self.seq_lens:
+            blocks_needed = (
+                int(math.ceil(seq_len / seq_stride)) if seq_stride > 0 else 0
+            )
+            row = []
+            for _ in range(blocks_needed):
+                row.append(self.parent.alloc_page())
+            self.seq_block_ids.append(row)
+
+    @property
+    def done(self) -> bool:
+        return len(self.done_result_indices) == self.bs
+
+    def detokenize(self) -> list[str]:
+        return self.parent.tokenizer.decode(self.results)
+
+    def print_current_results(self):
+        results = self.detokenize()
+        for i, s in enumerate(results):
+            seq_len = int(self.seq_lens[i])
+            print(f"  {i}({len(self.results[i])}, {seq_len}): {s}")
+
+    def add_result_token(self, tokens: torch.Tensor):
+        for i in range(self.bs):
+            token = tokens[i][0]
+            if token == self.parent.end_token:
+                self.done_result_indices.add(i)
+            if i in self.done_result_indices:
+                continue
+            token = int(tokens[i, 0])
+            self.results[i].append(token)
+
+    def allocate_seq_block_ids(self):
+        for i in range(self.bs):
+            sl = int(self.seq_lens[i])
+            if (sl % self.parent.block_seq_stride) == 0:
+                needed_blocks = sl // self.parent.block_seq_stride + 1
+            else:
+                needed_blocks = math.ceil(sl / self.parent.block_seq_stride)
+            block_ids_row = self.seq_block_ids[i]
+            while len(block_ids_row) < needed_blocks:
+                block_ids_row.append(self.parent.alloc_page())
+
+    def prefill(self):
+        model = self.parent.model
+        attention_mask = model.attention_mask(
+            model.input_mask(self.seq_lens, self.token_ids.shape[1])
+        )
+        seq_block_ids_tensor = self.pad_block_ids()
+        # print(f":: Invoke prefill:")
+        trace_tensor("prefill.token_ids", self.token_ids)
+        trace_tensor("prefill.seq_block_ids", seq_block_ids_tensor)
+        trace_tensor("prefill.attention_mask", attention_mask)
+        # print("prefill.token_ids", self.token_ids)
```

Comment:
It could be handy to have a `self.debug` that conditionally prints all of these. Obviously not necessary just would be handy.